apps/mark-scan/pat-device-input/src/patinputd.rs

```diff
@@ -0,0 +1,138 @@
+    device.click(&key)?;
+    device.synchronize().unwrap();
+    Ok(())
+}
+
+fn create_virtual_device() -> Device {
+    uinput::open(UINPUT_PATH)
+        .unwrap()
+        .name("PAT Input Daemon Virtual Device")
+        .unwrap()
+        .event(Keyboard::All)
+        .unwrap()
+        .create()
+        .unwrap()
+}
+
+fn set_up_pins() -> io::Result<()> {
+    IS_CONNECTED_PIN.set_up()?;
+    SIGNAL_A_PIN.set_up()?;
+    SIGNAL_B_PIN.set_up()?;
+    Ok(())
+}
+
+fn main() {
+    set_app_name(APP_NAME);
+    log!(
+        EventId::ProcessStarted;
+        EventType::SystemAction
+    );
+
+    let running = Arc::new(AtomicBool::new(true));
+
+    if let Err(e) = ctrlc::set_handler({
+        let running = running.clone();
+        move || {
+            running.store(false, Ordering::SeqCst);
+        }
+    }) {
+        log!(
+            event_id: EventId::ErrorSettingSigintHandler,
+            message: e.to_string(),
+            event_type: EventType::SystemStatus
+        );
+    }
+
+    // Create virtual device for keypress events
+    log!(
+          EventId::CreateVirtualUinputDeviceInit;
+          EventType::SystemAction
+    );
+    let mut device = create_virtual_device();
+    // Wait for virtual device to register
+    thread::sleep(Duration::from_secs(1));
+    log!(
+
+        event_id: EventId::CreateVirtualUinputDeviceComplete,
+        disposition: Disposition::Success,
+        event_type: EventType::SystemAction
+    );
+
+    log!(EventId::ConnectToPatInputInit);
+
+    // Export pins and set direction
+    set_up_pins().unwrap_or_else(|error| {
+        log!(
+            event_id: EventId::ConnectToPatInputComplete,
+            disposition: Disposition::Failure,
+            message: format!("An error occurred during GPIO pin connection: {error}")
+        );
+    });
+
+    // is_connected is unused for keypresses in this daemon but it's important to export
+    // the pin so it can be read by the mark-scan backend
+    let is_connected = IS_CONNECTED_PIN.is_active();
+    let mut signal_a = SIGNAL_A_PIN.is_active();
+    let mut signal_b = SIGNAL_B_PIN.is_active();
+
+    log!(
+        event_id: EventId::ConnectToPatInputInit,
+        message: format!("Connected to PAT with initial values [is_connected={is_connected}], [signal_a={signal_a}], [signal_b={signal_b}]")
+    );
+
+    loop {
+        if !running.load(Ordering::SeqCst) {
+            log!(
+                event_id: EventId::ProcessTerminated,
+                event_type: EventType::SystemAction
+            );
+            IS_CONNECTED_PIN.tear_down();
+            SIGNAL_A_PIN.tear_down();
+            SIGNAL_B_PIN.tear_down();
+            exit(0);
+        }
+
+        let new_signal_a = SIGNAL_A_PIN.is_active();
+        let new_signal_b = SIGNAL_B_PIN.is_active();
+
+        if new_signal_a && !signal_a {
+            send_key(&mut device, keyboard::Key::_1).unwrap();
```

Comment:
Do we want these to crash if it fails? Or just log an error?

Reply:
Initially I was thinking this should crash because it's useless without successful keypress events, but I suppose the errors could be intermittent. In that case we wouldn't want to crash. So far I haven't seen uinput fail in any way, so it's hard to say what type of error we'd get.

All said, it's probably safer to log the error and continue. Process restarts can always be achieved by restarting the machine.

Reply:
[779745a](https://github.com/votingworks/vxsuite/pull/4569/commits/779745a8a06502e8e4b235adce8509d2536ad1e9)